### PR TITLE
Optimize Projections

### PR DESCRIPTION
**Pull request by berkaysynnada**
## Which issue does this PR close?

Closes #.

## Are these changes tested?

```
┏━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Query        ┃     main ┃ feature_optimize_projections ┃        Change ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ QQuery 1     │ 421.93ms │                     421.60ms │     no change │
│ QQuery 2     │  59.86ms │                      59.83ms │     no change │
│ QQuery 3     │ 141.49ms │                     140.42ms │     no change │
│ QQuery 4     │  89.99ms │                      88.31ms │     no change │
│ QQuery 5     │ 173.94ms │                     171.95ms │     no change │
│ QQuery 6     │  84.61ms │                      84.30ms │     no change │
│ QQuery 7     │ 265.81ms │                     251.49ms │ +1.06x faster │
│ QQuery 8     │ 203.04ms │                     200.24ms │     no change │
│ QQuery 9     │ 285.64ms │                     285.41ms │     no change │
│ QQuery 10    │ 224.73ms │                     220.80ms │     no change │
│ QQuery 11    │  57.97ms │                      58.64ms │     no change │
│ QQuery 12    │ 158.46ms │                     158.17ms │     no change │
│ QQuery 13    │ 231.70ms │                     221.08ms │     no change │
│ QQuery 14    │ 119.61ms │                     119.42ms │     no change │
│ QQuery 15    │  82.82ms │                      83.33ms │     no change │
│ QQuery 16    │  51.07ms │                      50.90ms │     no change │
│ QQuery 17    │ 222.45ms │                     204.10ms │ +1.09x faster │
│ QQuery 18    │ 428.66ms │                     436.40ms │     no change │
│ QQuery 19    │ 239.63ms │                     241.73ms │     no change │
│ QQuery 20    │ 137.80ms │                     141.81ms │     no change │
│ QQuery 21    │ 354.14ms │                     362.21ms │     no change │
│ QQuery 22    │  52.01ms │                      53.07ms │     no change │
└──────────────┴──────────┴──────────────────────────────┴───────────────┘
```

### CODE REVIEW

File: datafusion/core/src/physical_optimizer/enforce_sorting.rs [hunk 1]
```diff
@@ -347,7 +351,12 @@ impl PhysicalOptimizerRule for EnforceSorting {
         // missed by the bottom-up traversal:
         let sort_pushdown = SortPushDown::init(updated_plan.plan);
         let adjusted = sort_pushdown.transform_down(&pushdown_sorts)?;
-        Ok(adjusted.plan)
+
+        let remove_projections = RemoveUnnecessaryProjections::init(adjusted.plan);
+        let projection_pushdown =
+            remove_projections.transform_down(&remove_unnecessary_projections)?;
+
```
Comment (mustafasrepo) [inline datafusion/core/src/physical_optimizer/enforce_sorting.rs:358]: I think this can be a rule itself. Not a sub-rule of EnforceSorting

File: datafusion/core/src/physical_optimizer/projection_pushdown.rs [hunk 2]
```diff
@@ -0,0 +1,2203 @@
...
+    if projection.benefits_from_input_partitioning()[0]
+        || physical_to_column_exprs(projection.expr()).is_none()
+    {
+        return Ok(Transformed::No(plans.clone()));
+    }
+
+    let new_projection =
+        projection_with_new_child(projection, repartition.input().clone())?;
+
+    let new_partitioning = match repartition.partitioning() {
```
Comment (mustafasrepo) [inline datafusion/core/src/physical_optimizer/projection_pushdown.rs:734]: Map hash expr also, as in sort, filter etc.

File: datafusion/physical-expr/src/scalar_function.rs [hunk 3]
```diff
@@ -109,6 +109,11 @@ impl ScalarFunctionExpr {
     pub fn return_type(&self) -> &DataType {
         &self.return_type
     }
+
+    /// Monotonicity information of the function
+    pub fn monotonicity(&self) -> &Option<FuncMonotonicity> {
```
Comment (metesynnada) [inline datafusion/physical-expr/src/scalar_function.rs:114]: Why is this included in this PR?

File: datafusion/core/src/physical_optimizer/projection_pushdown.rs [hunk 4]
```diff
@@ -0,0 +1,2073 @@
...
+}
+
+/// Tries to update the equi-join Column's of a join as if the projection replaced as the input of the join.
+fn update_join_on(
+    projection_left_exprs: &[(Column, String)],
+    projection_right_exprs: &[(Column, String)],
+    hash_join_on: &[(Column, Column)],
+) -> Option<Vec<(Column, Column)>> {
+    let (old_left_indices, old_right_indices): (Vec<Column>, Vec<Column>) =
+        hash_join_on.iter().cloned().unzip();
```
Comment (berkaysynnada) [inline datafusion/core/src/physical_optimizer/projection_pushdown.rs:1198]: How can we remove this clone?
